drivers/quota/projectquota.go

```diff
@@ -75,6 +78,25 @@ type Control struct {
 	quotas            map[string]uint32
 }
 
+// Attempt to generate a unigue projectid.  Multiple directories
+// per file system can have quota and they need a group of unigue
+// ids. This function attempts to allocate at least minPorjectIDS(10000)
+// unigue projectids, based on the inode of the basepath.
+func generateUnigueProjectID(path string) (uint32, error) {
+	const minProjectids = 10000
+	fileinfo, err := os.Stat(path)
+	if err != nil {
+		return 0, err
+	}
+	stat, ok := fileinfo.Sys().(*syscall.Stat_t)
+	if !ok {
+		return 0, fmt.Errorf("Not a syscall.Stat_t %s", path)
+
+	}
+	projectID := uint32(((stat.Ino * minProjectids) % math.MaxUint32))
```

Comment:
Are the project IDs supposed to be unique across the whole OS? (I have no idea, but the existence of `/etc/projects` and `/etc/projid` seems to suggest so.) If they are supposed to be OS-global, how does this ensure non-c/storage users don’t conflict?

---

There already is a `findNextProjectID` mechanism. Why does this do something so very different?

Reply:
At the very least, this should be something like `projectID := minProjectID + (stat.Ino * minProjectids) % (math.MaxUint32 - minProjectID)`, with `minProjectID` determined by the caller, so that this does not wrap and overlap the likely-allocated values [0, 1, 2, …] for those systems where the administrators intentionally set a high `minProjectID` by allocating one.

Reply:
Done

Reply:
Oops, `minProjectID` is not available to this function, that’s why it’s called in the first place. The current version looks reasonable.